### PR TITLE
feat: add `--root` cli option and make basic js api work

### PR DESCRIPTION
**Pull request by Timeless0911**
## Summary

- Add `--root` CLI param as base dir to load Rslib config file and .env file
- Use `config.root` which is default to `process.cwd()` as base dir to load `package.json` and `tsconfig.json`

close: #356 
close: #361 

## Related Links

#456 

## Checklist

- [ ] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

**Comment by Timeless0911** [comment]
@binarykitchen I think this PR can solve your issue.

File: packages/core/src/cli/build.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import type { RslibConfig } from '../types/config';
+import { getAbsolutePath } from '../utils/helper';
+import type { BuildOptions } from './commands';
+
+export async function build(
+  config: RslibConfig,
+  options: Pick<BuildOptions, 'root' | 'lib' | 'watch'> = {},
+): Promise<RsbuildInstance> {
+  const cwd = process.cwd();
+  options.root = options.root ? getAbsolutePath(cwd, options.root) : cwd;
```
Comment (fi3ework) [inline packages/core/src/cli/build.ts:12]: `options.root ` has is calculated in `packages/core/src/cli/init.ts#10` and passed to hereby. Do we need to re-calculate it again? (so does mf).

File: packages/core/src/cli/init.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import { loadConfig } from '../config';
+import type { RslibConfig } from '../types';
+import { getAbsolutePath } from '../utils/helper';
+import type { CommonOptions } from './commands';
+
+export async function getRslibConfig(
+  options: CommonOptions,
+): Promise<{ root: string; rslibConfig: RslibConfig }> {
+  const cwd = process.cwd();
+  const root = options.root ? getAbsolutePath(cwd, options.root) : cwd;
```
Comment (fi3ework) [inline packages/core/src/cli/init.ts:10]: IIUC, the path of config file does not take account into the `root` calculation anymore?

File: packages/core/src/cli/build.ts (hunk 1 above)
Comment (Timeless0911) [reply packages/core/src/cli/build.ts:12]: This `export async function build(
  config: RslibConfig,
  options: Pick<BuildOptions, 'root' | 'lib' | 'watch'> = {},
): Promise<RsbuildInstance>` function is aim to be our javascript api, so the absolute root path should be generated since `packages/core/src/cli/init.ts#10` is the cli logic. This file is put into the `cli` folder which may be not very good. Since we are not design the javascript api well, we can keep it here.

File: packages/core/src/cli/init.ts (hunk 2 above)
Comment (Timeless0911) [reply packages/core/src/cli/init.ts:10]: Yes, the `root` is as the base dir to find to config files.

File: packages/core/src/cli/build.ts (hunk 1 above)
Comment (fi3ework) [reply packages/core/src/cli/build.ts:12]: It's better not to modify parameter to avoid potential bug.

Comment (binarykitchen) [reply packages/core/src/cli/build.ts:12]: Agree. Anything async is very much prone to bugs. And like @fi3ework said, modifying parameters is a no-go.

Function parameters should remain **immutable**.


_Remember, UI = fn(data)_

File: packages/core/src/cli/inspect.ts
Comment (binarykitchen) [inline packages/core/src/cli/inspect.ts:1]: Wow, Webpack doesn't have this :)

Interesting ... just a little feedback, the function name `inspect` what does it truly inspect? If it changes configuration, I'd rename the function maybe?

File: packages/core/tests/__snapshots__/config.test.ts.snap [hunk 3]
```diff
@@ -72,7 +72,7 @@ exports[`Should compose create Rsbuild config correctly > Merge Rsbuild config 1
           "pnpapi",
         ],
         "filename": {
-          "js": "[name].js",
+          "js": "[name].mjs",
```
Comment (binarykitchen) [inline packages/core/tests/__snapshots__/config.test.ts.snap:75]: Why the change to mjs?

File: tests/integration/cli/build.test.ts
Comment (binarykitchen) [inline tests/integration/cli/build.test.ts:1]: Thanks for adding unit tests +++++++

File: packages/core/src/cli/commands.ts [hunk 4]
```diff
@@ -88,21 +91,13 @@ export function runCli(): void {
...
-          envMode: options.envMode,
-        });
-        const environments = await composeRsbuildEnvironments(rslibConfig);
-        const rsbuildInstance = await createRsbuild({
-          rsbuildConfig: {
-            environments: pruneEnvironments(environments, options.lib),
-          },
-        });
-        await rsbuildInstance.inspectConfig({
+        const { root, rslibConfig } = await getRslibConfig(options);
```
[the remark this answers is not in the document]
Comment (Timeless0911) [reply packages/core/src/cli/commands.ts:94]: Environment variables may not be passed in during runtime since user may use js api to start the child process without inherit some env. 

The reason for passing root is to read tsconfig and package.json. I think we can extract the relevant logic to the outside when we doing configuration verification and no configuration file mode in the future.

File: packages/core/tests/__snapshots__/config.test.ts.snap (hunk 3 above)
Comment (Timeless0911) [reply packages/core/tests/__snapshots__/config.test.ts.snap:75]: Because of the directory of the test case, package.json was not resolved successfully here just like your issue before, so `autoExtension` was not enabled.

File: packages/core/src/cli/inspect.ts
Comment (Timeless0911) [reply packages/core/src/cli/inspect.ts:1]: see https://lib.rsbuild.dev/guide/basic/cli#rslib-inspect, it does not changes configuration but print detailed configuration to better help users debug.

File: packages/core/tests/__snapshots__/config.test.ts.snap (hunk 3 above)
Comment (binarykitchen) [reply packages/core/tests/__snapshots__/config.test.ts.snap:75]: Ah, looks like this will fix more bugs than just one ...
... do we have a unit test case for this?

Comment (Timeless0911) [reply packages/core/tests/__snapshots__/config.test.ts.snap:75]: Current tests is enough since it correctly use the root package.json to generate extensions. This snapshot is used to check config merge behaviour, we do not find the issue of not finding package.json in this snapshot before. So this snapshot is just the unit test case for this.
 
And about autoExtension, there already have unit tests.

**Review by fi3ework — DISMISSED** [review]

File: packages/core/src/cli/commands.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (binarykitchen) [reply packages/core/src/cli/commands.ts:94]: > And also we will design our javascript API, so the root can be put into a context to pass around.

Ok, good, thanks for explaining. Do you already have a ticket for this?

[the remark this answers is not in the document]
Comment (Timeless0911) [reply packages/core/src/cli/commands.ts:94]: We are discussing that we should follow Rsbuild to resolve package.json and tsconfig.json, I think the package.json should be in folder as same as webpack context. https://rsbuild.dev/config/root#root

**Review by fi3ework — APPROVED** [review]

**Review by fi3ework — APPROVED** [review]
